tests/unit/test_charm.py

```diff
@@ -987,3 +994,351 @@ def test_given_ca_cert_is_not_stored_when_certificate_transfer_relation_joins_th
+        patch_create_snapshot.assert_called_once()
+
+    @patch("vault.Vault.is_sealed", new=Mock)
+    @patch("vault.Vault.unseal", new=Mock)
+    @patch("vault.Vault.is_initialized")
+    @patch("vault.Vault.create_snapshot")
+    def test_given_vault_is_not_initialized_when_create_raft_snapshot_then_snapshot_is_not_created(
+        self,
+        patch_create_snapshot,
+        patch_is_initialized,
+    ):
+        patch_is_initialized.return_value = False
+        self.harness.set_can_connect(container=self.container_name, val=True)
+        self.harness.add_storage(storage_name="certs", attach=True)
+        self.harness.charm._create_raft_snapshot()
+        patch_create_snapshot.assert_not_called()
+
+    @patch("vault.Vault.is_sealed", new=Mock)
+    @patch("vault.Vault.unseal", new=Mock)
+    @patch("vault.Vault.is_initialized")
+    @patch("vault.Vault.is_api_available")
+    @patch("vault.Vault.create_snapshot")
+    def test_given_vault_api_not_available_when_create_raft_snapshot_then_snapshot_is_not_created(
+        self,
+        patch_create_snapshot,
+        patch_is_api_available,
+        patch_is_initialized,
+    ):
+        patch_is_initialized.return_value = True
+        patch_is_api_available.return_value = False
+        self.harness.set_can_connect(container=self.container_name, val=True)
+        self.harness.add_storage(storage_name="certs", attach=True)
+        self.harness.charm._create_raft_snapshot()
+        patch_create_snapshot.assert_not_called()
+
+    @patch("vault.Vault.is_sealed", new=Mock)
+    @patch("vault.Vault.unseal", new=Mock)
+    @patch("vault.Vault.is_initialized")
+    @patch("vault.Vault.is_api_available")
+    @patch("vault.Vault.create_snapshot")
+    def test_given_vault_initialization_secret_not_available_when_create_raft_snapshot_then_snapshot_is_not_created(
+        self,
+        patch_create_snapshot,
+        patch_is_api_available,
+        patch_is_initialized,
+    ):
+        patch_is_initialized.return_value = True
+        patch_is_api_available.return_value = True
+        self.harness.set_can_connect(container=self.container_name, val=True)
+        self.harness.add_storage(storage_name="certs", attach=True)
+        self.harness.charm._create_raft_snapshot()
+        patch_create_snapshot.assert_not_called()
+
+    def test_given_cannot_connect_to_session_when_create_s3_bucket_then_exception_is_raised(self):
+        with self.assertRaises(Exception):
+            self.harness.charm._create_s3_bucket(
+                session=None,  # type: ignore[arg-type]
+                bucket_name="whatever bucket",
+                region="whatever region",
+                endpoint="whatever endpoint",
+            )
+
+    def test_given_bucket_already_exists_when_create_s3_bucket_then_bucket_not_created(
+        self,
+    ):
+        patch_session = Mock()
+        patch_session.resource.Bucket.meta.client.head_bucket.return_value = Mock()
```

Comment:
You shouldn't need this line, because a Mock returns a Mock for each property that is accessed.
```suggestion
```

Reply:
removed